libp2p/src/bandwidth.rs

```diff
@@ -209,3 +215,32 @@ impl<SMInner: AsyncWrite> AsyncWrite for InstrumentedStream<SMInner> {
         this.inner.poll_close(cx)
     }
 }
+
+// TODO: Ideally this should go somewhere else. I.e. good to not depend on prometheus-client in libp2p.
+pub fn register_bandwidth_sinks(
+    registry: &mut prometheus_client::registry::Registry,
+    sinks: Arc<RwLock<HashMap<String, Arc<BandwidthSinks>>>>,
+) {
+    registry.register_collector(Box::new(SinksCollector(sinks)));
+}
+
+#[derive(Debug)]
+struct SinksCollector(Arc<RwLock<HashMap<String, Arc<BandwidthSinks>>>>);
+
+impl prometheus_client::collector::Collector for SinksCollector {
+    fn encode(&self, mut encoder: DescriptorEncoder) -> Result<(), std::fmt::Error> {
+        let mut family_encoder =
+            encoder.encode_descriptor("bandwidth", "todo", None, MetricType::Counter)?;
+        for (protocols, sink) in self.0.read().expect("todo").iter() {
+            let labels = [("protocols", protocols.as_str()), ("direction", "inbound")];
+            let metric_encoder = family_encoder.encode_family(&labels)?;
+            ConstCounter::new(sink.inbound.load(Ordering::Relaxed)).encode(metric_encoder)?;
+
+            let labels = [("protocols", protocols.as_str()), ("direction", "outbound")];
+            let metric_encoder = family_encoder.encode_family(&labels)?;
+            ConstCounter::new(sink.outbound.load(Ordering::Relaxed)).encode(metric_encoder)?;
+        }
+
+        Ok(())
+    }
+}
```

Comment:
Where should this live?

- If in `libp2p`, the crate would thus need to import `prometheus-client`. I like the current approach where only `libp2p-metrics` depends on `prometheus-client`.
- If in `libp2p-metrics`, then we would have a circular dependency. I.e. `libp2p` would import `libp2p-metrics` in order to expose it via `libp2p::metrics`. `libp2p-metrics` would depend on `libp2p` in order to access `BandwidthSinks`.
- Yet another new crate?

@thomaseizinger do you have thoughts on this?

Reply:
Can't the entire bandwidth logging stuff live in `libp2p-metrics` and `libp2p` just provides the integration with the `SwarmBuilder`?

i.e. `libp2p-metrics` can re-export the `Registry` so you can reference it in the type signatures. We already pull `prometheus-client` into the public API as part of `libp2p-metrics` anyway. But if you re-export, you don't need explicitly declare a dependency.

Reply:
Sounds good. I should have a commit ready later today.